log4j2/src/main/java/com/google/common/flogger/backend/log4j2/ValueList.java

```diff
@@ -0,0 +1,156 @@
+            ValueList existingList = existingValue instanceof ValueList ? (ValueList) existingValue
+                    : new ValueList(existingValue, null);
+            return new ValueList(value, existingList);
+        }
+    }
+
+    @Override
+    public Iterator<Object> iterator() {
+        return new LinkedIterator(this);
+    }
+
+    static void appendValues(String label, Object valueOrList, MetadataKey.KeyValueHandler kvh) {
+        if (valueOrList instanceof ValueList) {
+            for (Object value : (ValueList) valueOrList) {
+                emit(label, value, kvh);
+            }
+        } else {
+            emit(label, valueOrList, kvh);
+        }
+    }
+
+    /**
+     * Tags are key-value mappings which cannot be modified or replaced. If you add
+     * the tag mapping "foo" -> true and later add the mapping "foo" -> false, you
+     * get the value "foo" mapped to both true and false. This is *very* deliberate
+     * since the key space for tags is global and the risk of two bits of code
+     * accidentally using the same tag name is real (e.g. you get "id=abcd" but you
+     * added "id=xyz" so you think this isn't your log entry, but someone else added
+     * "id=abcd" in a context you weren't aware of).
+     *
+     * Given three tag mappings: "baz" -> , "foo" -> true and "foo" -> false the
+     * Value list is going to flatten the tags in order to store them as tags=[baz,
+     * foo=false, foo=true].
+     * 
+     * Reusing the label 'tags' is intentional as this allows us to store the
+     * flatten tags in Log4j2's ContextMap.
+     */
+    static void emit(String label, Object value, MetadataKey.KeyValueHandler kvh) {
+        if (value instanceof Tags) {
+            // Flatten tags to treat them as keys or key/value pairs, e.g. tags=[foo,
+            // baz=bar2, baz=bar]
+            ((Tags) value).asMap().forEach((k, v) -> {
+                if (v.isEmpty()) {
+                    kvh.handle(label, k);
+                } else {
+                    for (Object obj : v) {
+                        kvh.handle(label, k + "=" + obj.toString());
+                    }
+                }
+            });
+        } else {
+            kvh.handle(label, value);
+        }
+    }
+
+    /**
+     * Returns a string representation of the contents of the specified value list.
+     * If the list contains other lists as elements, they are converted to strings.
+     *
+     * <p>
+     * If value list is empty, this method returns {@code "null"}.
+     * <p>
+     * If value list contains a single element {@code a}, this method returns
+     * {@code "a"}.
+     * <p>
+     * Otherwise the content of teh list is returned surrounded by brackets "[", "].
+     */
+    @Override
+    public String toString() {
+        if (value == null) {
+            return "null";
+        }
+
+        if (next == null || next.value == null) {
```

Comment:
Feels like one of these should not happen. If "EMPTY" is always present then "next.value == null" should suffice (since value is only null in the sentinel). However my preference is actually to not bother with EMPTY, use "null" as the end-of-list and make the value non-nullable in the constructor. That way "I have a list node" ==> "its value is non-null", which feels like a neater thing to say. It also lets you not need to worry about "value == null" above.

Reply:
I replaced the value list with a queue.